### ISSUE TITLE
[Roadmap] Distributed Linear Models

### ISSUE THREAD

**Issue by CodingCat** [dmlc/xgboost#4203]
We have been seeing the requirement to have an efficient distributed linear model implementation which would be used either for better model explanation or ensembling with tree model

We are going to work on this

**Comment by thvasilo** [comment]
One question I had more generally is how does linear model training fit into the purpose of XGBoost overall.

Does the community feel like the project should be specialized to GBRTs, given that tools like VowpalWabbit are perfectly fine for scalable linear model training?

**Comment by CodingCat** [comment]
Eh...there have been linear model in xgboost though single machine supported only

**Comment by tqchen** [comment]
I agree that XGBoost should focus on trees. 

The reason that there was a linear model is because linear model also fit into the paradigm of boosting and we could sometime ensemble linear model with a tree directly, so it is not a priority, but if there is some requirements to get a ensemble version, it is also not bad to have

**Comment by CodingCat** [comment]
one of the important requirements I am seeing in business side is , if X changes with 0.1, how many influences it will bring to the output Y

this type of explanation is critical in business operating side, and it is hard to achieve with a tree model


if XGB can provide a one-stop solution, it will further pushes the usage of XGB

**Comment by thvasilo** [comment]
Yes, I think this is a design decision that the community needs to make. If XGBoost if focused on trees and boosting then it should do that and it well (always aim to be the fastest and most accurate). That allows the codebase to be focused and optimized for tree-based learning.

If the community decides to further support linear models (which haven't gotten much attention since they were released in XGB AFAIK) we might be moving XGB from its original purpose into a more general ML framework. I'd argue there are enough libraries out there to perform distributed linear learning, from the aforementioned VW, to all major ML frameworks (TF, MXNet, PyTorch), and SparkML to mention a few.

Of course, following the Apache merit-based way of decision making, my opinion shouldn't drive a decision, but i thought more senior community members would want to chime in before delving into something like this. It's up to the community to decide where to take XGB feature-wise.

**Comment by CodingCat** [comment]
I don’t think “others already did” is a good reason to prevent something happening in this case

For instance, why there is MxNet/TF/PyTorch? and even TF can do some kind of tree learning

Regarding the code complexity, the current architecture has been able to support a new updater within gradient boosting framework, not to mention, this is more about adding param syncing calls properly in an existing updater

**Comment by CodingCat** [comment]
BTW, the last major updates to linear model is about GPU-based coordinate descent which was added last April, 

and...(I don’t want to list how many features we have added to tree models just to prove the community is not distracted)

**Comment by trivialfis** [comment]
> For instance, why there is MxNet/TF/PyTorch

Honestly,  I still want to reinvent the wheel...

@CodingCat I'm interested in other models than trees.  Let me know if I can help in any ways.

**Comment by trivialfis** [comment]
I don't think we will implement this in foreseeable future.
